# Выбор локации для скважины

In [1]:
import warnings
warnings.filterwarnings("ignore") 
import numpy as np 
import pandas as pd 
import scipy.stats

from sklearn.model_selection import train_test_split 

import math 
from sklearn.linear_model import LinearRegression  
from sklearn.metrics import mean_squared_error 
from numpy.random import RandomState 

In [2]:
df_0=pd.read_csv('/datasets/geo_data_0.csv') 
df_0.info()
display(df_0.head(10))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647
5  wX4Hy  0.969570  0.489775 -0.735383   64.741541
6  tL6pL  0.645075  0.530656  1.780266   49.055285
7  BYPU6 -0.400648  0.808337 -5.624670   72.943292
8  j9Oui  0.643105 -0.551583  2.372141  113.356160
9  OLuZU  2.173381  0.563698  9.441852  127.910945

In [3]:
print('Доля пропусков в столбце ,%:') 
rezul=df_0.isna().mean() * 100
print(rezul)
print('Дубликатов = ',df_0.duplicated().sum()) 

Доля пропусков в столбце ,%:
id         0.0
f0         0.0
f1         0.0
f2         0.0
product    0.0
dtype: float64
Дубликатов =  0


In [4]:
df_1=pd.read_csv('/datasets/geo_data_1.csv') 
df_1.info()
display(df_1.head(10))


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305
5  HHckp  -3.327590  -2.205276  3.003647   84.038886
6  h5Ujo -11.142655 -10.133399  4.002382  110.992147
7  muH9x   4.234715  -0.001354  2.004588   53.906522
8  YiRkx  13.355129  -0.332068  4.998647  134.766305
9  jG6Gi   1.069227 -11.025667  4.997844  137.945408

In [5]:
print('Доля пропусков в столбце ,%:') 
rezul=df_1.isna().mean() * 100
print(rezul)
print('Дубликатов = ',df_1.duplicated().sum()) 

Доля пропусков в столбце ,%:
id         0.0
f0         0.0
f1         0.0
f2         0.0
product    0.0
dtype: float64
Дубликатов =  0


In [6]:
df_2=pd.read_csv('/datasets/geo_data_2.csv') 
df_2.info()
display(df_2.head(10))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746
5  LzZXx -0.758092  0.710691  2.585887   90.222465
6  WBHRv -0.574891  0.317727  1.773745   45.641478
7  XO8fn -1.906649 -2.458350 -0.177097   72.480640
8  ybmQ5  1.776292 -0.279356  3.004156  106.616832
9  OilcN -1.214452 -0.439314  5.922514   52.954532

In [7]:
print('Доля пропусков в столбце ,%:') 
rezul=df_2.isna().mean() * 100
print(rezul)
print('Дубликатов = ',df_2.duplicated().sum()) 

Доля пропусков в столбце ,%:
id         0.0
f0         0.0
f1         0.0
f2         0.0
product    0.0
dtype: float64
Дубликатов =  0


In [8]:
df_0 = df_0.drop(['id'], axis=1) 
df_1 = df_1.drop(['id'], axis=1) 
df_2 = df_2.drop(['id'], axis=1) 

## Обучение и проверка модели

In [9]:
features_0 =df_0.drop(['product',], axis=1) 
target_0 =df_0['product']  
target_0 = pd.Series(target_0) 

features_1 =df_1.drop(['product',], axis=1) 
target_1 =df_1['product']  
target_1 = pd.Series(target_1) 

features_2 =df_2.drop(['product',], axis=1) 
target_2 =df_2['product']  
target_2 = pd.Series(target_2) 

# разделяем на выборку для обучения и тестирования, 25% данных для валидации (теста)
features_train_0, features_valid_0, target_train_0, target_valid_0 = train_test_split(features_0, target_0, test_size=0.25, random_state=2504)
features_train_1, features_valid_1, target_train_1, target_valid_1 = train_test_split(features_1, target_1, test_size=0.25, random_state=2504)
features_train_2, features_valid_2, target_train_2, target_valid_2 = train_test_split(features_2, target_2, test_size=0.25, random_state=2504)

In [10]:
model_0 = LinearRegression() 
model_0.fit(features_train_0, target_train_0) 
predictions_0 = model_0.predict(features_valid_0)
result_0=math.sqrt(mean_squared_error( target_valid_0, predictions_0,squared=True)) 
print('Средний запас сырья в регионе 0 =', predictions_0.mean())
print('RMSE =', result_0) 

Средний запас сырья в регионе 0 = 92.5536068080489
RMSE = 37.68797312271404


In [11]:
model_1 = LinearRegression() 
model_1.fit(features_train_1, target_train_1) 
predictions_1 = model_1.predict(features_valid_1)
result_1 =math.sqrt(mean_squared_error(target_valid_1, predictions_1,squared=True)) 
print('Средний запас сырья в регионе 1 =', predictions_1.mean())
print('RMSE =', result_1) 

Средний запас сырья в регионе 1 = 69.22428166460364
RMSE = 0.8963537624826772


In [12]:
model_2  = LinearRegression() 
model_2.fit(features_train_2, target_train_2) 
predictions_2 = model_2.predict(features_valid_2)
result_2 =math.sqrt(mean_squared_error(target_valid_2, predictions_2,squared=True))
print('Средний запас сырья в регионе 2 =', predictions_2.mean())
print('RMSE', result_2) 

Средний запас сырья в регионе 2 = 94.94685045023044
RMSE 39.86197791046957


**Вывод**
Наименьшая ошибка предсказания получилась для региона 1, но в данном регионе наименьшее количество запасов нефти. Наибольшее количество запасов в регионах 0 и 2, но ошибка предсказания значительно больше. Для выбора региона необходимо оценить прибыль и риски.

## Подготовка к расчёту прибыли

Достаточный объём сырья для безубыточной разработки новой скважины

In [13]:
TOTAL_MANY=pow(10,10) 
PRICE_OIL=4.5*pow(10,5) 
BOREHOLE=200 
Q=TOTAL_MANY/(PRICE_OIL*BOREHOLE) 
print('Минимальный объем месторождений в регионе', Q)

Минимальный объем месторождений в регионе 111.11111111111111


**Вывод**
Во всех регионах среднее значение объемов нефти меньше минимального значения для окупаемовсти, но необходимо рассмотреть и выбрать 200 месторождений со значением выше 111 тыс. барелей.

## Расчёт прибыли и рисков 

In [14]:
# Функция расчета дохада от предсказаний и по геологоразведке
def Out_many (Q , features, target, model_pr):  
    summa_df=0
    summa_predict=0 
    spisok_skvagin=[]
    
    data_pr = model_pr.predict(features)
    data_pr=pd.Series(data_pr) 
    data_pr_f=data_pr.sort_values(ascending=False) 
    data_pr_f= data_pr_f[0:BOREHOLE] 
    summa_predict= data_pr_f.sum()*PRICE_OIL 

    spisok_skvagin=target[data_pr_f.index] 
    summa_df=spisok_skvagin.sum()*PRICE_OIL   
    
    return  summa_predict, summa_df,data_pr

### Доход и прибиль в регионе 0

In [15]:
many_predict, many_df,data_pedict_0=Out_many(Q,features_0,target_0,model_0)
print('Регион 0')
print(f'Доход с наилучших 200 скважин по предсказанию модели {many_predict/pow(10,9):.4} млрд. руб.')
print(f'Доход с тех же 200 скважин, по результатам геологоразведки {many_df/pow(10,9):.4} млрд. руб.')
print(f'Прибль с тех же 200 скважин {(many_predict-many_df)/pow(10,9):.4} млрд. руб.')
print(f'Разница предсказания модели по отношекнию к геологоразведке составляет {abs((many_predict-many_df)/many_df)*100:.3} %')

Регион 0
Доход с наилучших 200 скважин по предсказанию модели 14.69 млрд. руб.
Доход с тех же 200 скважин, по результатам геологоразведки 13.49 млрд. руб.
Прибль с тех же 200 скважин 1.198 млрд. руб.
Разница предсказания модели по отношекнию к геологоразведке составляет 8.88 %


### Доход и прибиль в регионе 1

In [16]:
many_predict,many_df,data_pedict_1=Out_many(Q,features_1,target_1,model_1)
print('Регион 1')
print(f'Доход с наилучших 200 скважин по предсказанию модели {many_predict/pow(10,9):.4} млрд. руб.')
print(f'Доход с тех же 200 скважин, по результатам геологоразведки {many_df/pow(10,9):.4} млрд. руб.')
print(f'Прибль с тех же 200 скважин  {(many_predict-many_df)/pow(10,9):.4} млрд. руб.')
print(f'Разница предсказания модели по отношекнию к геологоразведке составляет  {abs((many_predict-many_df)/many_df)*100:.3} %')

Регион 1
Доход с наилучших 200 скважин по предсказанию модели 12.52 млрд. руб.
Доход с тех же 200 скважин, по результатам геологоразведки 12.42 млрд. руб.
Прибль с тех же 200 скважин  0.1092 млрд. руб.
Разница предсказания модели по отношекнию к геологоразведке составляет  0.88 %


### Доход и прибиль в регионе 2

In [17]:
many_predict,many_df,data_pedict_2=Out_many(Q,features_2,target_2,model_2)
print('Регион 2')
print(f'Доход с наилучших 200 скважин по предсказанию модели {many_predict/pow(10,9):.4} млрд. руб.')
print(f'Доход с тех же 200 скважин, по результатам геологоразведки {many_df/pow(10,9):.4} млрд. руб.')
print(f'Прибль с тех же 200 скважин {(many_predict-many_df)/pow(10,9)}:.4 млрд. руб.')
print(f'Разница предсказания модели по отношекнию к геологоразведке составляет {abs((many_predict-many_df)/many_df)*100:.3} %')

Регион 2
Доход с наилучших 200 скважин по предсказанию модели 14.09 млрд. руб.
Доход с тех же 200 скважин, по результатам геологоразведки 12.62 млрд. руб.
Прибль с тех же 200 скважин 1.4710423460156155:.4 млрд. руб.
Разница предсказания модели по отношекнию к геологоразведке составляет 11.7 %


### Процедура *Bootstrap*

In [18]:

def revenue(target, predictions, wells):
    predictions_sorted = predictions.sort_values(ascending=False)
    selected = target[predictions_sorted.index][:wells].sum() 
    return (selected*PRICE_OIL -TOTAL_MANY)

#### Регион 0

In [33]:
# функиция посчнта параметров для анализа региона
def score (target,data_pedict,number): 
    state = np.random.RandomState(2504)
    values = []
    coun_positiv=0
    bootstrap_samples = 1000
    for i in range(bootstrap_samples):
        target_sub_sample = target.sample(n=500, random_state=state,replace=True)
        predict_sub_sample=data_pedict[target_sub_sample.index] 
        profit = revenue(target_sub_sample, predict_sub_sample, 200)/pow(10,9)
        if profit>0:
            coun_positiv+=1
        values.append(profit) 
    values = pd.Series(values)
    lower =values.quantile(0.025)
    upper =values.quantile(0.975)
    print('Для региона',number)
    print('Средняя прибль',values.mean(),'млрд руб')
    print(f'95%-ый доверительный интервал {lower:.4} , {upper:.4} млрд руб')
    print('Вероятность убытков',(1-(coun_positiv/len(values)))*100,'%')

In [34]:
score(target_0,data_pedict_0,0)

Для региона 0
Средняя прибль 0.428507952928524 млрд руб
95%-ый доверительный интервал -0.08182 , 0.9331 млрд руб
Вероятность убытков 5.200000000000005 %


#### Регион 1

In [35]:
score(target_1,data_pedict_1,1)

Для региона 1
Средняя прибль 0.46812449998383326 млрд руб
95%-ый доверительный интервал 0.07034 , 0.9086 млрд руб
Вероятность убытков 1.100000000000001 %


#### Регион 2

In [36]:
score(target_2,data_pedict_2,2)

Для региона 2
Средняя прибль 0.3725070551973921 млрд руб
95%-ый доверительный интервал -0.1808 , 0.8471 млрд руб
Вероятность убытков 8.099999999999996 %


## **Вывод**
По полученным результатам видно, что вероятность убытков меньше всего в регионе 1. В регионах 0 и 2 вероятность убытков значительно выше. Так же в регионе 1 модель предсказания имеет наименьшую ошибку. Стоит обратить внимание, что в регионах 0 и 2 потенциальная прибль выше, но так же значительно выше риски.